src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -174,16 +174,13 @@ private Object readResolve() {
         execution = CpsFlowExecution.PROGRAM_STATE_SERIALIZATION.get();
         setupTransients();
         assert execution!=null;
-        if (scripts != null) { // compatibility: the field will be null in old programs
+        if (scripts != null && !scripts.isEmpty()) { // compatibility: the field will be null in old programs
             GroovyShell shell = execution.getShell();
-            assert shell.getContext().getVariables().isEmpty();
-            if (!scripts.isEmpty()) {
-                // Take the canonical bindings from the main script and relink that object with that of the shell and all other loaded scripts which kept the same bindings.
-                shell.getContext().getVariables().putAll(scripts.get(0).getBinding().getVariables());
-                for (Script script : scripts) {
-                    if (script.getBinding().getVariables().equals(shell.getContext().getVariables())) {
-                        script.setBinding(shell.getContext());
-                    }
+            // Take the canonical bindings from the main script and relink that object with that of the shell and all other loaded scripts which kept the same bindings.
+            shell.getContext().getVariables().putAll(scripts.get(0).getBinding().getVariables());
```

Comment:
So this will overwrite any variable bindings present in the original shell. Is that what you meant to do?

Reply:
> it would probably be better only overwrite variables that are not present in the context already

Not sure. I _think_ the unconditional overwrite as currently implemented is fine; just wanted to bring up the question. Ideally the test would pin a specific behavior.

Reply:
Added a test `clashingVariablesOnRestart` that shows the unconditional overwrite behavior